Source/controls/keymapper.hpp

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <string>
+#include <functional>
+#include <vector>
+#include <unordered_map>
+
+namespace devilution {
+
+// The Keymapper maps keys to actions.
+class Keymapper final {
+public:
+	using SetConfigKeyFunction = std::function<void(const std::string &key, const std::string &value)>;
+	using GetConfigKeyFunction = std::function<std::string(const std::string &key)>;
+	using ActionIndex = std::size_t;
+
+	// Action represents an action that can be triggered using a keyboard
+	// shortcut.
+	class Action final {
+	public:
+		// Can this action be triggered while the player is dead?
+		enum class IfDead {
+			Allow,
+			Ignore,
+		};
+
+		Action(const std::string &name, int defaultKey, std::function<void()> action)
+		    : name(name)
+		    , defaultKey(defaultKey)
+		    , action(action)
+		    , ifDead(IfDead::Ignore)
+		{
+		}
+		Action(const std::string &name, int defaultKey, std::function<void()> action, IfDead ifDead)
+		    : name(name)
+		    , defaultKey(defaultKey)
+		    , action(action)
+		    , ifDead(ifDead)
+		{
+		}
+
+		void operator()() const
+		{
+			action();
+		}
+
+	private:
+		std::string name;
+		int defaultKey;
+		std::function<void()> action;
+		IfDead ifDead;
+		int key {};
+
+		friend class Keymapper;
+	};
+
+	// Keymapper, for now, uses two function pointers to interact with the
+	// configuration.
+	// This is mostly a workaround and should be replaced later when another INI
+	// library will be used.

```

Comment:
```suggestion
	/**
	 * Keymapper, for now, uses two function pointers to interact with the
	 * configuration.
	 * This is mostly a workaround and should be replaced later when another INI
	 * library will be used.
	 */
```

Reply:
Why use C style comments? I usually prefer C++ ones because you can quickly comment out part of your code, and if it's full of C style /* */ you can't. Doxygen support [both](https://www.doxygen.nl/manual/docblocks.html).

Reply:
Yeah I also prefer C++ comments, fewer lines